src/Dialect/ONNX/DisposableElementsAttributeStorage.hpp

```diff
@@ -0,0 +1,118 @@
+
+  static DisposableElementsAttributeStorage *construct(
+      AttributeStorageAllocator &allocator, const KeyTy &key) {
+    ShapedType type = std::get<TYPE>(key);
+    Strides strides = std::get<STRIDES>(key);
+    onnx_mlir::BType bufferBType = std::get<BUFFER_BTYPE>(key);
+    onnx_mlir::BType btype = std::get<BTYPE>(key);
+    bool isContiguous = std::get<IS_CONTIGUOUS>(key);
+    size_t id = std::get<ID>(key);
+    return new (allocator.allocate<DisposableElementsAttributeStorage>())
+        DisposableElementsAttributeStorage(type, allocator.copyInto(strides),
+            bufferBType, btype, isContiguous, id);
+  }
+
+  // The tensor shape and element type that this object represents.
+  // The underlying data in buffer may not directly match the type's element
+  // type or number of elements, depending on strides and reader.
+  ShapedType type;
+
+  // Specifies how to map positions expressed in type's shape to the flat
+  // indices in buffer. strides can express that buffer is not in the default
+  // row-major order (maybe as a result of a transpose) or requires broadcast
+  // to fill in type's shape. A special case is when the buffer holds a single
+  // splat value that broadcasts to shape's size with all-zero strides.
+  Strides strides;
+
+  // Data type of the elements in buffer before transform.
+  onnx_mlir::BType bufferBType;
+
+  // Data type (BOOL, INT8, FLOAT16, etc) of the type's elements.
+  // Redundant as btype == btypeOfMlirType(type.getElementType())
+  // but we store it for fast access.
+  onnx_mlir::BType btype;
+
+  // Do the strides match the type's shape?
+  // Redundant as isContiguous == areStridesContiguous(type.getShape(), strides)
+  // but we store it for fast access.
+  bool isContiguous;
+
+  // Serial number that distinguishes instances.
+  size_t id;
+
+  // shared_ptr to an underlying MemoryBuffer which can be either heap allocated
+  // or a mmap'ed file or point to the raw data of a DenseElementsAttr.
+  //
+  // The buffer elements' data type may not match type's element type, namely
+  // when the transform function transforms the buffer data type to another
+  // data type.
+  //
+  // Garbage collection clears the buffer when the DisposableElementsAttr is
+  // disposed.
+  //
+  // Multiple DisposableElementsAttr can point to the same MemoryBuffer.
+  // The MemoryBuffer is destroyed (and heap allocated data freed or mmap'ed
+  // file closed) when no one points to it anymore.
+  Buffer buffer;
+
+  // Reads the buffer elements to WideNums corresponding to type's
+  // element type. Is null if data is not transformed.
+  // In this case the buffer data type and the type's element type must promote
+  // to the same double/i64/u64 widetype
+  // (both are float, or both are signed ints, or both are unsigned ints).
+  //
+  // Garbage collection clears the reader when the DisposableElementsAttr is
+  // disposed.
+  Reader reader;
+};
+
+} // namespace mlir
```

Comment:
not that it matters a lot, but it looks like the formatter like empty line at the end of the file.

Reply:
done